File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        l, r = 0, len(matrix) - 1

        while l <= r:
            m = (l + r) // 2
            left, right = 0, len(matrix[m]) - 1
            
            while left <= right:
                mid = (left + right) // 2

                if target < matrix[m][left]:
                    r = m - 1
                    break
                    
                if target > matrix[m][right]:
                    l = m + 1
                    break

                if target > matrix[m][mid]:
                    left = mid + 1
                elif target < matrix[m][mid]:
                    right = mid - 1
                else:
                    return True
        
        return False

File: Data Structures & Algorithms/is-palindrome/submission-1.java
class Solution {
    public boolean isPalindrome(String s) {
        String str = s.replace(" ", "").toLowerCase();

        int left = 0;
        int right = str.length() - 1;

        while (left < right) {
            while (left < right && !isAlphaNum(str.charAt(left), str)) left++;
            while (left < right && !isAlphaNum(str.charAt(right), str)) right--;
            
            if (str.charAt(left) != str.charAt(right)) return false;

            left++;
            right--;
        }

        return true;
    }

    public boolean isAlphaNum(char name, String str) {
        return Character.isLetter(name) || Character.isDigit(name);
    }
}
